relay-cardinality/src/redis.rs

```diff
@@ -50,171 +49,256 @@ impl RedisSetLimiter {
-                        entries,
-                        timestamp,
-                        limit,
-                    )
-                }
-            )?;
+            let results = metric!(timer(CardinalityLimiterTimers::Redis), id = state.id, {
+                self.check_limits(&mut connection, &mut state, timestamp)
+            })?;
 
             // This always acquires a write lock, but we only hit this
             // if we previously didn't satisfy the request from the cache,
             // -> there is a very high chance we actually need the lock.
-            let mut cache = self
-                .cache
-                .update(organization_id, &item_scope, cache_window);
-
+            let mut cache = self.cache.update(state.scope, timestamp); // Acquire a write lock.
             for (entry, status) in results {
                 if status.is_rejected() {
-                    acceptions_rejections.miss(&item_scope);
                     rejected.push(entry.rejection());
+                    state.rejected();
                 } else {
-                    acceptions_rejections.hit(&item_scope);
                     cache.accept(entry.hash);
+                    state.accepted();
                 }
             }
+            drop(cache); // Give up the cache lock!
         }
 
         if !rejected.is_empty() {
             relay_log::debug!(
-                organization_id = organization_id,
+                scoping = ?scoping,
                 "rejected {} metrics due to cardinality limit",
                 rejected.len(),
             );
         }
 
-        acceptions_rejections.report(
-            CardinalityLimiterCounters::Accepted,
-            CardinalityLimiterCounters::Rejected,
-        );
-
         Ok(Box::new(rejected.into_iter()))
     }
 }
 
+/// A quota scoping extracted from a [`CardinalityLimit`] and a [`Scoping`].
+#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq)]
+pub(crate) struct QuotaScoping {
+    pub window: SlidingWindow,
+    pub namespace: Option<MetricNamespace>,
+    pub organization_id: Option<OrganizationId>,
+    pub project_id: Option<ProjectId>,
+}
+
+impl QuotaScoping {
+    /// Creates a new [`QuotaScoping`] from a [`Scoping`] and [`CardinalityLimit`].
+    ///
+    /// Returns `None` for limits with scope [`CardinalityScope::Unknown`].
+    pub fn new(scoping: Scoping, limit: &CardinalityLimit) -> Option<Self> {
+        let (organization_id, project_id) = match limit.scope {
+            CardinalityScope::Organization => (Some(scoping.organization_id), None),
+            // Invalid/unknown scope -> ignore the limit.
+            CardinalityScope::Unknown => return None,
+        };
+
+        Some(Self {
+            window: limit.window,
+            namespace: limit.namespace,
+            organization_id,
+            project_id,
+        })
+    }
+
+    /// Wether the scoping applies to the passed entry.
+    pub fn applies_to(&self, entry: &Entry) -> bool {
```

Comment:
nit: Similar functions are called `matches` elsewhere. Up to you whether you want to align it. Example: https://github.com/getsentry/relay/blob/2d7aa88d4c2251c428e7cf891ff8c9e7ebf131f1/relay-quotas/src/rate_limit.rs#L170-L171